### ISSUE TITLE
Wine has problems when built with GCC 9

### ISSUE THREAD

**Issue by TheGreatMcPain** [InBetweenNames/gentooLTO#346]
There are few problems with wine when built with GCC 9.

Such as EA Origin not being able to go online, and more listed here.
https://bugs.winehq.org/show_bug.cgi?id=47167

I know this isn't really an issue related to GentooLTO, but I thought I'd give people a heads up before they upgrade GCC. Well at least until wine gets fixed.

**Comment by funghetto** [comment]
I believe you can upgrade and keep the old GCC just for wine.

**Comment by pchome** [comment]
It's fun how following lines was changed in the provided logs:
```log
0034:fixme:winhttp:get_system_proxy_autoconfig_url no support on this platform
0034:fixme:winhttp:WinHttpDetectAutoProxyConfigUrl discovery via DHCP not supported
003a:err:virtual:map_file_into_view failed to set PROT_EXEC on file map, noexec filesystem?
003a:err:virtual:virtual_map_section mapping 0xfb60000 80000 000000000 failed
003a:err:virtual:map_file_into_view failed to set PROT_EXEC on file map, noexec filesystem?
003a:err:virtual:virtual_map_section mapping 0xfb60000 80000 000000000 failed
009c:fixme:kernelbase:AppPolicyGetThreadInitializationType 0xfffffffa, 0x1130fea0
```
```log
0034:fixme:winhttp:get_system_proxy_autoconfig_url no support on this platform
0034:fixme:winhttp:WinHttpDetectAutoProxyConfigUrl discovery via DHCP not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\028d" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0202" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\028c" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\01f6" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0998\8000" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\02b2" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\02fd" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\020e" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0210" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0291" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0222" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0998\8000" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0293" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0243" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\02c7" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\026d" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0298" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\024a" not supported
0031:fixme:bcrypt:BCryptOpenAlgorithmProvider algorithm L"\0388" not supported
009c:fixme:kernelbase:AppPolicyGetThreadInitializationType 0xfffffffa, 0x1098fea0
```
[map_file_into_view](https://source.winehq.org/git/wine.git/blob/48a74277f5f9e0c87d8c01457236a89d94cb71d3:/dlls/ntdll/virtual.c#l1138)

I'd say GCC 9 fixes things, filesystem now accessible and "exec". ;)

**Comment by pchome** [comment]
I compiled standalone kernel32 test (`virtual` part) to check if there any errors connected to `map_*`. The interesting part of log (GCC 9), but TBH I'm not sure it's related:
```log
virtual.c:3955: Test failed: 5: CreateFileMapping(0020) should fail
virtual.c:3956: Test failed: expected ERROR_ACCESS_DENIED, got 0
virtual.c:3955: Test failed: 7: CreateFileMapping(0080) should fail
virtual.c:3956: Test failed: expected ERROR_ACCESS_DENIED, got 0
```
[dlls/kernel32/tests/virtual.c](https://source.winehq.org/git/wine.git/blob/130c9b34455ac50811b80184b94124eaffc63139..2ee8554a0ae53597234c2b5fa6a95124d2dedeab:/dlls/kernel32/tests/virtual.c#l3892)

For those who want to play with this:
[test-kernel32.zip](https://github.com/InBetweenNames/gentooLTO/files/3269715/test-kernel32.zip)

_NOTE: "staging" - just a marker, to run different Wine versions in different prefixes and produce separate log files._
_NOTE: `WINETEST_PLATFORM=windows` to disable "todo" markers_
_[How the test was built (example)](https://github.com/pchome/wine-playground/tree/master/wine-proton-tests)_

**Comment by ionenwks** [comment]
Looking at https://bugs.winehq.org/show_bug.cgi?id=47167 again, there's now a patch for it at https://source.winehq.org/patches/data/166316

Seems the problem wasn't with gcc9 but rather older version of gcc that were allowing incorrect behavior (which wine ended up relying on). Hopefully will get fixed in 4.11, for now I just threw a user patch to my wine (planning to uninstall gcc 8.3 soon so I was looking at loose ends).

**Comment by pchome** [comment]
I guess we should use something like `-Werror=c90-c99-compat` with gcc9.

**Comment by pchome** [comment]
https://source.winehq.org/git/wine.git/commitdiff/9afc341c4f043240f0dc3de6351550b03d24b131

**Comment by FireBurn** [comment]
Thanks, this fixed quite a few issues I'd been having lately - I'd assumed it was a change in bcrypt / gnutls rather than with GCC

**Comment by funghetto** [comment]
This should be closed, the fix landed in wine 4.11

**Comment by bobwya** [comment]
For anyone that cares... I've backported this fix to the all the older Wine versions I support, in my [Github: bobwya/bobwya](https://github.com/bobwya/bobwya) Overlay...

**Comment by pchome** [comment]
@bobwya 
Just curious, why gentoo and overlay owners still using *.msi in mono ebuilds?

I using shared mono builds in my local overlay since supported (4.8.x ?) w/o any problems. Maybe I missed something? Any known bugs or restrictions?

<details>
  <summary>[app-emulation/wine-mono ebuild diff]</summary>

```
--- a/app-emulation/wine-mono/wine-mono-4.9.0.ebuild
+++b/app-emulation/wine-mono/wine-mono-4.9.0.ebuild
@@ -5,7 +5,7 @@
 
 DESCRIPTION="Wine Mono is a replacement for the .NET runtime and class libraries in Wine"
 HOMEPAGE="https://www.winehq.org/"
-SRC_URI="https://dl.winehq.org/wine/${PN}/${PV}/${P}.msi"
+SRC_URI="https://dl.winehq.org/wine/${PN}/${PV}/${PN}-bin-${PV}.tar.gz"
 
 LICENSE="BSD-2 GPL-2 LGPL-2.1 MIT MPL-1.1"
 SLOT="${PV}"
@@ -17,5 +18,5 @@
 
 src_install() {
        insinto /usr/share/wine/mono
-       doins "${DISTDIR}/${P}.msi"
+        doins -r ${P}
 }
```
</details>

<details>
  <summary>[What]</summary>

Shared mono installstion. https://www.winehq.org/pipermail/wine-devel/2019-April/143379.html
> With yesterday's Git update, Wine now has the ability to use a shared
install of Wine Mono. This greatly reduces the disk space needed by
each prefix, and it probably speeds up prefix creation as well. Those
changes will be in Wine 4.6.
</details>

**Comment by bobwya** [comment]
@pchome 

On my part it's laziness. I'm well aware of the changes Upstream. I'm always fiddling about with something "more important"... :-)

As for Gentoo itself... There is a bit of a vacuum, in the maintenance of the Gentoo Wine ebuilds, since NP stepped down.

**Comment by ionenwks** [comment]
Been waiting for that shared mono support as well, out of laziness (as well), I haven't bothered to set it up yet in my local overlay while hoping that gentoo does it for me, but at this rate I guess I might as well.

**Comment by InBetweenNames** [comment]
Closing as the offending code was fixed in Wine 4.11.  Thanks!
